In [1]:
#importamos primero seaborn y el df planetas 

import seaborn as sb

planets = sb.load_dataset("planets")


In [3]:
# analizamos el df

planets.shape

(1035, 6)

In [5]:
planets.head(10)

method  number  orbital_period   mass  distance  year
0  Radial Velocity       1         269.300   7.10     77.40  2006
1  Radial Velocity       1         874.774   2.21     56.95  2008
2  Radial Velocity       1         763.000   2.60     19.84  2011
3  Radial Velocity       1         326.030  19.40    110.62  2007
4  Radial Velocity       1         516.220  10.50    119.47  2009
5  Radial Velocity       1         185.840   4.80     76.39  2008
6  Radial Velocity       1        1773.400   4.64     18.15  2002
7  Radial Velocity       1         798.500    NaN     21.41  1996
8  Radial Velocity       1         993.300  10.30     73.10  2008
9  Radial Velocity       2         452.800   1.99     74.79  2010

In [7]:
planets.describe()  # como summary en R

number  orbital_period        mass     distance         year
count  1035.000000      992.000000  513.000000   808.000000  1035.000000
mean      1.785507     2002.917596    2.638161   264.069282  2009.070531
std       1.240976    26014.728304    3.818617   733.116493     3.972567
min       1.000000        0.090706    0.003600     1.350000  1989.000000
25%       1.000000        5.442540    0.229000    32.560000  2007.000000
50%       1.000000       39.979500    1.260000    55.250000  2010.000000
75%       2.000000      526.005000    3.040000   178.500000  2012.000000
max       7.000000   730000.000000   25.000000  8500.000000  2014.000000

Vemos que en count, tenemos cuantos datos hay, ecomo hay 1035 registros, habra NA ya que mass por ejemplo solo tiene 513 observaciones, para ello eliminamos dichos NA

In [9]:
planets.dropna().describe()

number  orbital_period        mass    distance         year
count  498.00000      498.000000  498.000000  498.000000   498.000000
mean     1.73494      835.778671    2.509320   52.068213  2007.377510
std      1.17572     1469.128259    3.636274   46.596041     4.167284
min      1.00000        1.328300    0.003600    1.350000  1989.000000
25%      1.00000       38.272250    0.212500   24.497500  2005.000000
50%      1.00000      357.000000    1.245000   39.940000  2009.000000
75%      2.00000      999.600000    2.867500   59.332500  2011.000000
max      6.00000    17337.500000   25.000000  354.000000  2014.000000

In [12]:
#vemos que drop na no borro datos
planets.describe()

number  orbital_period        mass     distance         year
count  1035.000000      992.000000  513.000000   808.000000  1035.000000
mean      1.785507     2002.917596    2.638161   264.069282  2009.070531
std       1.240976    26014.728304    3.818617   733.116493     3.972567
min       1.000000        0.090706    0.003600     1.350000  1989.000000
25%       1.000000        5.442540    0.229000    32.560000  2007.000000
50%       1.000000       39.979500    1.260000    55.250000  2010.000000
75%       2.000000      526.005000    3.040000   178.500000  2012.000000
max       7.000000   730000.000000   25.000000  8500.000000  2014.000000

In [10]:
# accedemos a var distancia y realizamos algunas operaciones
planets.distance



0        77.40
1        56.95
2        19.84
3       110.62
4       119.47
         ...  
1030    172.00
1031    148.00
1032    174.00
1033    293.00
1034    260.00
Name: distance, Length: 1035, dtype: float64

In [11]:
planets.distance.sum()


213367.98000000004

In [14]:
planets.distance.prod()

/home/miguel/anaconda3/lib/python3.11/site-packages/numpy/core/_methods.py:53: RuntimeWarning: overflow encountered in reduce
  return umr_prod(a, axis, dtype, out, keepdims, initial, where)


inf

lo siguiente es la desvicion media absoluta
$$\frac {  \sum_{n=1}^n |x_i-\bar{x}|       }  {n}$$

In [16]:
# se calculaba con
import pandas as ps
planets.distance.mad()

# pero pandas ya no existe

AttributeError: 'Series' object has no attribute 'mad'

In [47]:
# creemos la funcion
def mad(x):
    d = abs(x - x.mean()).mean()
    return d



In [49]:
mad(planets.distance)

318.10088545240666

In [51]:
planets.distance.var()

537459.7922212934

In [52]:
planets.distance.std()

733.1164929404422

 El .groupby en python es similar al aggregate o by, donde variable(s)~factor(es) solo que de la forma
 
df.groupby('factor(es)')['columna(s)'].FUN
 

In [53]:
planets.groupby('method')["orbital_period"].median()

method
Astrometry                         631.180000
Eclipse Timing Variations         4343.500000
Imaging                          27500.000000
Microlensing                      3300.000000
Orbital Brightness Modulation        0.342887
Pulsar Timing                       66.541900
Pulsation Timing Variations       1170.000000
Radial Velocity                    360.200000
Transit                              5.714932
Transit Timing Variations           57.011000
Name: orbital_period, dtype: float64

In [55]:
#otros ejemplos
planets.groupby('method')[["orbital_period",'year']].median()


orbital_period    year
method                                               
Astrometry                         631.180000  2011.5
Eclipse Timing Variations         4343.500000  2010.0
Imaging                          27500.000000  2009.0
Microlensing                      3300.000000  2010.0
Orbital Brightness Modulation        0.342887  2011.0
Pulsar Timing                       66.541900  1994.0
Pulsation Timing Variations       1170.000000  2007.0
Radial Velocity                    360.200000  2009.0
Transit                              5.714932  2012.0
Transit Timing Variations           57.011000  2012.5

In [57]:
#otros ejemplos
planets.groupby(['year','method'])[["orbital_period"]].median().unstack()


orbital_period  ...                          
method     Astrometry  ... Transit Timing Variations
year                   ...                          
1989              NaN  ...                       NaN
1992              NaN  ...                       NaN
1994              NaN  ...                       NaN
1995              NaN  ...                       NaN
1996              NaN  ...                       NaN
1997              NaN  ...                       NaN
1998              NaN  ...                       NaN
1999              NaN  ...                       NaN
2000              NaN  ...                       NaN
2001              NaN  ...                       NaN
2002              NaN  ...                       NaN
2003              NaN  ...                       NaN
2004              NaN  ...                       NaN
2005              NaN  ...                       NaN
2006              NaN  ...                       NaN
2007              NaN  ...                       NaN
2008              NaN  ...                       NaN
2009              NaN  ...                       NaN
2010          1016.00  ...                       NaN
2011              NaN  ...                  160.0000
2012              NaN  ...                   57.0110
2013           246.36  ...                   22.3395
2014              NaN  ...                       NaN

[23 rows x 10 columns]

# creando factores
supongamos que queremos abrupar por decada, para ello primero hay que crear dicha variable, tomando año, lo asignaremos a una decada




In [73]:
decada = planets.year//10 *10
decada

0       2000
1       2000
2       2010
3       2000
4       2000
        ... 
1030    2000
1031    2000
1032    2000
1033    2000
1034    2000
Name: year, Length: 1035, dtype: int64

In [66]:
planets.groupby([decada,'method'])[["orbital_period"]].sum().unstack()

orbital_period  ...                          
method     Astrometry  ... Transit Timing Variations
year                   ...                          
1980              NaN  ...                       NaN
1990              NaN  ...                       NaN
2000              NaN  ...                       NaN
2010          1262.36  ...                  239.3505

[4 rows x 10 columns]

In [74]:
# incluso podriamos hacerlo como coloquialmente se hace con "80s"
decada = decada.astype(str) +'s'
decada

0       2000s
1       2000s
2       2010s
3       2000s
4       2000s
        ...  
1030    2000s
1031    2000s
1032    2000s
1033    2000s
1034    2000s
Name: year, Length: 1035, dtype: object

In [76]:
planets.groupby([decada,'method'])[["orbital_period"]].sum().unstack()

orbital_period  ...                          
method     Astrometry  ... Transit Timing Variations
year                   ...                          
1980s             NaN  ...                       NaN
1990s             NaN  ...                       NaN
2000s             NaN  ...                       NaN
2010s         1262.36  ...                  239.3505

[4 rows x 10 columns]

In [77]:
# notemos que la va decada no tiene mas que el nombre heredado year, por tanto podemos cambiarlo con
decada.name = "decada"
planets.groupby([decada,'method'])[["orbital_period"]].sum().unstack()

orbital_period  ...                          
method     Astrometry  ... Transit Timing Variations
decada                 ...                          
1980s             NaN  ...                       NaN
1990s             NaN  ...                       NaN
2000s             NaN  ...                       NaN
2010s         1262.36  ...                  239.3505

[4 rows x 10 columns]

In [79]:
# quitemos los na
planets.groupby([decada,'method'])[["number"]].sum().unstack().fillna(0)

number                            ...                                  
method Astrometry Eclipse Timing Variations  ... Transit Transit Timing Variations
decada                                       ...                                  
1980s         0.0                       0.0  ...     0.0                       0.0
1990s         0.0                       0.0  ...     0.0                       0.0
2000s         0.0                       5.0  ...    64.0                       0.0
2010s         2.0                      10.0  ...   712.0                       9.0

[4 rows x 10 columns]